Electrified component

spookydonut opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Adds a component for airlocks and vending machines to use for their electrification behaviour since its essentially a carbon copy for both.
Can be expanded to apply to other types of machines in future.

This has been tested and appears to work.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Lets badmins electrify any machine they want.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
code: Anything can be electrified now with electrification being a component
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

ninjanomnom commented on code/datums/wires/airlock.dm:
```diff
@@ -18,7 +19,7 @@
 /datum/wires/airlock/interactable(mob/user)
 	var/obj/machinery/door/airlock/A = holder
-	if(!issilicon(user) && A.isElectrified() && A.shock(user, 100))
+	if(!issilicon(user) && (SEND_SIGNAL(A, COMSIG_AIRLOCK_WIRES_INTERACT, user) & COMSIG_ELECTRIFIED_SHOCK))
```
you should not have side effects in condition check functions like this

spookydonut replied:
I'll make a different PR to fix this, how should it be done?

spookydonut replied:
Disregard

ninjanomnom commented on code/datums/components/electrified.dm:
```diff
@@ -0,0 +1,143 @@
+			return
+		duration_timer = addtimer(CALLBACK(src, .proc/end_electrification), duration, TIMER_STOPPABLE|TIMER_OVERRIDE)
+
+	RegisterSignal(src, COMSIG_ELECTRIFICATION_REMOVAL, .proc/end_electrification)
```
This is what GetComponent is for, put removal behaviour in Destroy

spookydonut replied:
It's for the stuff where it toggles electrification, this signal is for that stuff to either <disable the electrification> and get a return value and if not then add electrification

spookydonut replied:
So should that use like
```
var/a = GetComponent(electrification)
if(a)
    qdel(a)
else
    AddComponent(electrification)
```
?

optimumtact commented on code/datums/components/electrified.dm:
```diff
@@ -0,0 +1,153 @@
+
+	ui_update()
+
+	RegisterSignal(parent, COMSIG_MACHINERY_POWERRESTORED, .proc/power_restored)
```
RegisterWithParent

and UnregisterFromParent

optimumtact commented on code/datums/components/electrified.dm:
```diff
@@ -0,0 +1,153 @@
+		addtimer(CALLBACK(src, .proc/end_electrification), duration, TIMER_STOPPABLE|TIMER_OVERRIDE)
+		duration = null
+
+/datum/component/electrified/proc/registersignals()
```
RegisterWithParent

and UnregisterFromParent

optimumtact replied:
scrap this, just call these from those procs

optimumtact commented on code/__DEFINES/components.dm:
```diff
@@ -213,6 +213,23 @@
+#define COMSIG_MACHINERY_BROKEN "machinery_broken"
+
+#define COMSIG_AIRLOCK_BUMPOPEN "airlock_bumpopen"					//from airlock bumpopen()
+	#define COMSIG_ELECTRIFIED_SHOCK 2							// COMPONENT_NO_ATTACK_HAND is 1
```
something went wrong

optimumtact commented on code/__DEFINES/components.dm:
```diff
@@ -213,6 +213,23 @@
+#define COMSIG_MACHINERY_BROKEN "machinery_broken"
+
+#define COMSIG_AIRLOCK_BUMPOPEN "airlock_bumpopen"					//from airlock bumpopen()
+	#define COMSIG_ELECTRIFIED_SHOCK 1
```
what is going on

spookydonut replied:
\shrug

Qustinnus replied:
Why is it a number

Qustinnus replied:
Nevermind I see

ninjanomnom replied:
It's a bitflag return value for the above signal, this is the normal format in this file

spookydonut committed:
progress

progress

progress

progress

progress

progress

fix this garbage

it feels yucky using getcomponent

vending machines

perm toggle

wait no

autodoc

this can be 1

ooops

merge fix

last fixes

AnturK commented on code/__DEFINES/components.dm:
```diff
@@ -213,11 +213,22 @@
+#define COMSIG_AIRLOCK_ATTACKBY_POWERCROWBAR "airlock_attackby_powercrowbar"
+#define COMSIG_AIRLOCK_ATTACK_ALIEN "airlock_attack_alien"
+#define COMSIG_ELECTRIFICATION_CHANGE "electrification_change"
+#define COMSIG_AIRLOCK_AI_ALTCLICK "airlock_ai_altclick"
```
Signal names referencing their singe use are icky. Name this something like AIRLOCK_SILICON_ELECTRIFY

ninjanomnom replied:
You have this backwards, signals should be named after what's happening to cause it, not what the signal should do

AnturK commented on code/datums/components/electrified.dm:
```diff
@@ -0,0 +1,279 @@
+GLOBAL_LIST_INIT(unelectrifiable_machines, typecacheof(list(/obj/machinery/door/airlock/cult)))
```
Why does this need to exist ? Wouldn't this make more sense as can electrify signal ?

spookydonut committed:
changes per review

spookydonut committed:
arghs

spookydonut committed:
appeasement

spookydonut committed:
look at me now dad

ninjanomnom commented on code/_onclick/ai.dm:
```diff
@@ -147,7 +147,7 @@
 		return
 
-	if(!secondsElectrified)
+	if(!SEND_SIGNAL(src, COMSIG_AIRLOCK_SILICON_ELECTRIFY))
```
Flag, not bool

ninjanomnom replied:
This should be moved to the base of AIAltClick and make this call parent

ninjanomnom commented on code/datums/components/electrified.dm:
```diff
@@ -0,0 +1,277 @@
+	log_electrification(src, user)
+
+	if(electrification_state == MACHINE_NOT_ELECTRIFIED)
+		return INITIALIZE_HINT_QDEL
```
```suggestion
		return COMPONENT_INCOMPATIBLE
```

ninjanomnom commented on code/datums/components/electrified.dm:
```diff
@@ -0,0 +1,277 @@
+		COMSIG_MOVABLE_IMPACT))
+
+/datum/component/electrified/machinery/register_interactions()
+	return
```
???

ninjanomnom commented on code/datums/components/electrified.dm:
```diff
@@ -0,0 +1,277 @@
+  *
+  */
+/datum/component/electrified/machinery/proc/electrified()
+	return 1
```
Use flag defines

ninjanomnom commented on code/datums/components/electrified.dm:
```diff
@@ -0,0 +1,277 @@
+	unregister_interactions()
+	return ..()
+
+/datum/component/electrified/machinery/Destroy(force, silent)
```
Dont intermix two different types like this

ninjanomnom commented on code/game/objects/obj_defense.dm:
```diff
@@ -104,6 +104,7 @@
 		playsound(src.loc, 'sound/weapons/slash.ogg', 100, TRUE)
 
 /obj/attack_animal(mob/living/simple_animal/M)
+	SEND_SIGNAL(src, COMSIG_OBJ_ATTACK_ANIMAL, M)
```
Put the signal at the base proc and call parent here

ninjanomnom commented on code/game/machinery/doors/airlock_types.dm:
```diff
@@ -446,17 +446,14 @@
 /obj/machinery/door/airlock/cult/Initialize()
 	. = ..()
 	new openingoverlaytype(loc)
+	RegisterSignal(src, COMSIG_CAN_ELECTRIFY, .proc/dont_electrify)
```
Don't add the component in the first place, this is a poor signal

ninjanomnom commented on code/game/machinery/doors/airlock.dm:
```diff
@@ -1006,9 +965,8 @@
-			if(isElectrified())
-				shock(user,100)//it's like sticking a forck in a power socket
-				return
+			if(SEND_SIGNAL(src, COMSIG_AIRLOCK_ATTACKBY_POWERCROWBAR, user) & COMSIG_ELECTRIFIED_SHOCK)
```
This signal should be sent regardless of state of the airlock, let the component decide if they get shocked

ninjanomnom commented on code/game/machinery/doors/airlock.dm:
```diff
@@ -799,10 +758,10 @@
-			if(shock(user, 75))
-				return
+	if(prob(75))
+		if(SEND_SIGNAL(src, COMSIG_AIRLOCK_ATTACKBY_CONDUCTIVE, user) & COMSIG_ELECTRIFIED_SHOCK)
```
1. The component should be deciding when it shocks
2. The signal should be sent every time if you're going to have this
3. Call parent and use the base attackby signal instead

actioninja commented:
stale

spookydonut commented:
> 
> 
> stale

probably need to start over on this
